### PR TITLE
[SDK-609]: Create Yoti v2 module

### PR DESCRIPTION
**Pull request by echarrod**
I've gone with the "Major branch" strategy outlined in the [Go wiki on Modules ](https://github.com/golang/go/wiki/Modules#releasing-modules-v2-or-higher). This allows us to carry on working on master, and if we do need to change anything for version 2 (after we've released version 3), then we can make a branch for v2, and make any bug fixes there. 
I'll add further detail to a wiki shortly, including what we'll need to do with each minor and major release going forwards.

When we release the major version, it will just be a case of renaming the module, and changing the v2s to v3s (or using [this handy tool](https://github.com/marwan-at-work/mod)). 

Also updated some fields to make them read only. 

This will require an update to the developer docs.

### CODE REVIEW

File: yoti_test.go [hunk 1]
```diff
@@ -1018,24 +1018,6 @@ func TestAnchorParser_YotiAdmin(t *testing.T) {
 	AssertServerCertSerialNo(t, "256616937783084706710155170893983549581", actualSerialNo)
 }
 
-func TestAnchors_Unknown(t *testing.T) {
```
Comment (echarrod) [inline yoti_test.go]: Since anchor type is unexported now, we can't create this unknown anchor in the way that we're doing in this test. I spoke with Mike and have created https://lampkicking.atlassian.net/browse/SDK-686 to reimplement this test once we have a mocking framework. Drayton also mentioned Core had some testing tools, which might solve this.

**Comment by echarrod** [comment]
Accompanying decision log: https://lampkicking.atlassian.net/wiki/spaces/SDK/pages/917831708/Version+Management+for+Go+SDK

**Comment by pn-santos** [comment]
Regarding the modules part, LGTM :+1:

File: yoti_test.go [hunk 2]
```diff
@@ -904,12 +904,12 @@ func TestAnchorParser_Passport(t *testing.T) {
 		t.Error("Anchors and Sources should be the same when there is only one Source")
 	}
 
-	if actualAnchor.Type != anchor.AnchorTypeSource {
-		t.Errorf("Parsed anchor type is incorrect. Expected: %q, actual: %q", anchor.AnchorTypeSource, actualAnchor.Type)
+	if actualAnchor.Type() != anchor.AnchorTypeSource {
+		t.Errorf("Parsed anchor type is incorrect. Expected: %q, actual: %q", anchor.AnchorTypeSource, actualAnchor.Type())
```
Comment (sjones303) [inline yoti_test.go]: Consider splitting long lines, currently have to scroll sideways on GitHub to view the whole line in quite a few places.

File: README.md [hunk 3]
```diff
@@ -56,15 +56,15 @@ Yoti also allows you to enable user details verification from your mobile app by
 To download and install the Yoti SDK and its dependencies, simply run the following command from your terminal:
 
 ```Go
-go get "github.com/getyoti/yoti-go-sdk"
+go get "github.com/getyoti/yoti-go-sdk/v2"
```
Comment (sjones303) [inline README.md]: Consider mentioning modules are now used, so it's not necessary get a copy or fetch all dependencies as the Go toolchain will fetch them as necessary. Unless someone wishes to run the examples or make a contribution it's sufficient to just start using it/add a `require` to `go.mod` and the toolchain will take care of the rest.

File: activitydetails.go [hunk 4]
```diff
@@ -0,0 +1,14 @@
...
+
+//ActivityDetails represents the result of an activity between a user and the application
+type ActivityDetails struct {
+	UserProfile  Profile
+	rememberMeID string
+}
+
+// RememberMeID is a unique identifier Yoti assigns to your user, but only for your app
+// if the same user logs into your app again, you get the same id
+// if she/he logs into another application, Yoti will assign a different id for that app
```
[the remark this answers is not in the document]
Comment (echarrod) [reply activitydetails.go]: Thanks, makes sense. I'll change to either have a blank line, or make sure they are separated with a full stop.

File: anchor/anchorparser.go [hunk 5]
```diff
@@ -28,18 +28,20 @@ func ParseAnchors(protoAnchors []*yotiprotoattr.Anchor) []*Anchor {
 
 		for _, cert := range parsedCerts {
 			for _, ext := range cert.Extensions {
-				var ae anchorExtension
 				if sourceOID.Equal(ext.Id) {
 					anchorType = AnchorTypeSource
+
+					extension := unmarshalExtension(ext.Value)
+					if extension != "" {
+						extensions = append(extensions, extension)
```
[the remark this answers is not in the document]
Comment (sjones303) [reply anchor/anchorparser.go]: That is in essence what I'm saying.

It's a shame the documentation from the protobufs has been stripped, the documentation I've seen clearly states the first certificate in the chain will have the source/verifier extension, this means if all we're interested in is the source/verifier (i.e. not verifying the chain) then we only need to process/parse the first certificate.

In addition to that a certificate may have many extensions (e.g. key use), but the leaf certificate will only have one Yoti source or one Yoti verifier extension, it may have other non-Yoti extensions. Therefore to find the anchor type and the source/verifier value we just need to loop through the extensions until a Yoti OID is found.

This code currently works because there is only ever one source or one verifier extension in the chain. If it were to ever process a malformed anchor (i.e. multiple source/verifier extensions) then the values slice will have multiple entries, but the type will only be of the last extension processed. This isn't too much of an issue if the anchor was trying to express multiple verifications (type would be verification, with multiple values), but if there was a mixture then both the type and values would be meaningless.

File: examples/go.mod [hunk 6]
```diff
@@ -0,0 +1,3 @@
+module github.com/getyoti/yoti-go-sdk/examples
+
+require github.com/joho/godotenv v1.3.0
```
[the remark this answers is not in the document]
Comment (echarrod) [reply examples/go.mod]: So it it best to leave this as is, and then we add it as a requirement once v2.3.0 is tagged?

File: anchor/anchorparser.go [hunk 7]
```diff
@@ -2,45 +2,47 @@ package anchor
...
-					extensions = append(extensions, ae.Extension)
-				} else {
-					log.Printf("Error unmarshalling anchor extension: %q", err)
+				var (
+					value string
+					err   error
+				)
+				anchorType, value, err = parseExtension(ext)
+				if err != nil {
+					log.Printf("error parsing anchor extension, %v", err)
```
Comment (sjones303) [inline anchor/anchorparser.go]: I know I suggested this, but I've realised we will now log an error for other extensions, which is incorrect. Maybe it would be better to change the `default` case in `parseExtension` to be:
```go
default:
	return anchorType, "", nil
```
Then you can do:
```go
if err != nil {
	⋮
} else if anchorType == AnchorTypeUnknown {
	continue
}
⋮
```

File: anchor/anchorparser.go [hunk 8]
```diff
@@ -52,6 +54,45 @@ func ParseAnchors(protoAnchors []*yotiprotoattr.Anchor) []*Anchor {
...
+		anchorType = AnchorTypeSource
+	case ext.Id.Equal(verifierOID):
+		anchorType = AnchorTypeVerifier
+	default:
+		return anchorType, "", errors.New("unknown anchor type")
+	}
+
+	var ae anchorExtension
+	_, err = asn1.Unmarshal(ext.Value, &ae)
+
```
Comment (sjones303) [inline anchor/anchorparser.go]: It generally make things easier to follow if there's no space between the function call and the error handling.

File: yoti_test.go [hunk 9]
```diff
@@ -217,7 +215,8 @@ func TestYotiClient_ParseWithoutProfile_Success(t *testing.T) {
 			result = &httpResponse{
 				Success:    true,
 				StatusCode: 200,
-				Content:    `{"receipt":{"wrapped_receipt_key": "` + wrappedReceiptKey + `",` + otherPartyProfileContent + `"remember_me_id":"` + rememberMeID + `", "sharing_outcome":"SUCCESS"}}`}
+				Content: `{"receipt":{"wrapped_receipt_key": "` + wrappedReceiptKey + `",` +
```
Comment (sjones303) [inline yoti_test.go]: Has this been run through `goimports`? It looks like the alignment is off.

File: yoti_test.go [hunk 10]
```diff
@@ -12,16 +12,17 @@ import (
...
 const sdkID = "fake-sdk-id"
+const wrappedReceiptKey = "kyHPjq2+Y48cx+9yS/XzmW09jVUylSdhbP+3Q9Tc9p6bCEnyfa8vj38AIu744RzzE+Dc4qkSF21VfzQKtJVILfOXu5xRc7MYa5k3zWhjiesg/gsrv7J4wDyyBpHIJB8TWXnubYMbSYQJjlsfwyxE9kGe0YI08pRo2Tiht0bfR5Z/YrhAk4UBvjp84D+oyug/1mtGhKphA4vgPhQ9/y2wcInYxju7Q6yzOsXGaRUXR38Tn2YmY9OBgjxiTnhoYJFP1X9YJkHeWMW0vxF1RHxgIVrpf7oRzdY1nq28qzRg5+wC7cjRpS2i/CKUAo0oVG4pbpXsaFhaTewStVC7UFtA77JHb3EnF4HcSWMnK5FM7GGkL9MMXQenh11NZHKPWXpux0nLZ6/vwffXZfsiyTIcFL/NajGN8C/hnNBljoQ+B3fzWbjcq5ueUOPwARZ1y38W83UwMynzkud/iEdHLaZIu4qUCRkfSxJg7Dc+O9/BdiffkOn2GyFmNjVeq754DCUypxzMkjYxokedN84nK13OU4afVyC7t5DDxAK/MqAc69NCBRLqMi5f8BMeOZfMcSWPGC9a2Qu8VgG125TuZT4+wIykUhGyj3Bb2/fdPsxwuKFR+E0uqs0ZKvcv1tkNRRtKYBqTacgGK9Yoehg12cyLrITLdjU1fmIDn4/vrhztN5w="
```
Comment (sjones303) [inline yoti_test.go]: Consider placing these in a single `const (…)` block, in my opinion it is cleaner, and can allow the whole block to be documented if necessary.

File: examples/go.mod (hunk 6 above)
[the remark this answers is not in the document]
Comment (sjones303) [reply examples/go.mod]: Yes, that's right.

File: yoti_test.go (hunk 9 above)
Comment (echarrod) [reply yoti_test.go]: I currently go through the steps on the [CONTRIBUTING.md](https://github.com/getyoti/yoti-go-sdk/blob/master/CONTRIBUTING.md) page, as well as doing `gometalinter`, so we're doing `go fmt`. Would you suggest replacing `go fmt` with `goimports`? 

In any case, this shouldn't have been missed by `go fmt`, doesn't seem to be changed when I run it now either. I'll investigate

File: anchor/anchorparser.go (hunk 7 above)
Comment (echarrod) [reply anchor/anchorparser.go]: Yes that's sensible, will change

File: anchor/anchorparser.go (hunk 8 above)
Comment (echarrod) [reply anchor/anchorparser.go]: Sure. Is it just your preference on using switch statements here instead of:
```
if err != nil {
    return anchorType, "", fmt.Errorf("unable to unmarshal extension: %v", err)
} else if len(ae.Extension) == 0 {
    return anchorType, "", errors.New("empty extension")
}

val = ae.Extension
```

I've always used the above when we're not switching on a definitive object type

File: yoti_test.go (hunk 10 above)
Comment (echarrod) [reply yoti_test.go]: Yep I agree, that would be cleaner

File: yoti_test.go (hunk 9 above)
Comment (sjones303) [reply yoti_test.go]: Yeah, I had a play too. Since `goimports` essentially wraps `go fmt` it has the same behaviour, wasn't aware of this edge case.

It would be worth some time in the future looking into using `goimports`, it does all that `go fmt` does, but also sanitises imports (e.g. removes unused ones to prevent build failures down the line).

File: anchor/anchorparser.go (hunk 8 above)
Comment (sjones303) [reply anchor/anchorparser.go]: They're both pretty much the same (I've not had a dig to see if they are optimised differently). It's a fairly common pattern to use a `switch` for more complex error (or just return values) processing, in my opinion since the logic can be expressed as scenarios (i.e. cases) it's cleaner to write it that way, rather than chaining `if`s. Unless the processing code is hard to follow as written I won't tend to suggest changing between `if`/`switch`.

**Comment by echarrod** [comment]
I have reverted the addition of modules for this minor release. We will look at adding them in the next release.

File: attribute/stringattribute.go [hunk 11]
```diff
@@ -5,14 +5,14 @@ import (
...
 type StringAttribute struct {
 	*yotiprotoattr.Attribute
 	value   string
 	anchors []*anchor.Anchor
 }
 
-//NewString creates a new String attribute
+// NewString creates a new String attribute
 func NewString(a *yotiprotoattr.Attribute) *StringAttribute {
 	parsedAnchors := anchor.ParseAnchors(a.Anchors)
```
Comment (sidibos) [inline attribute/stringattribute.go]: Not sure about this approach, maybe the parser shoddy have been in a different module. Like AnchorParser or Converter.

File: anchor/anchorparser.go [hunk 12]
```diff
@@ -14,33 +17,34 @@ type anchorExtension struct {
...
-					log.Printf("Error unmarshalling anchor extension: %q", err)
+				var (
+					value string
+					err   error
+				)
+				anchorType, value, err = parseExtension(ext)
+				if err != nil {
+					log.Printf("error parsing anchor extension, %v", err)
+					continue
+				} else if anchorType == AnchorTypeUnknown {
```
[the remark this answers is not in the document]
Comment (echarrod) [reply anchor/anchorparser.go]: We will be parsing the unknown anchors, but we won't be parsing the extensions on those anchors. See the discussion on SDK-683

File: attribute/stringattribute.go (hunk 11 above)
Comment (echarrod) [reply attribute/stringattribute.go]: the `anchor` at the start is a package qualifier - i.e. we are accessing the `ParseAnchors` function in the `anchor` package.

Comment (sidibos) [reply attribute/stringattribute.go]: 👍

File: anchor/anchorparser.go (hunk 12 above)
[the remark this answers is not in the document]
Comment (sidibos) [reply anchor/anchorparser.go]: 👍
